apm-agent-plugins/apm-hibernate-search-plugin/apm-hibernate-search-plugin-5_x/src/main/java/co/elastic/apm/agent/hibernate/search/v5_x/HibernateSearch5Instrumentation.java

```diff
@@ -0,0 +1,111 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.hibernate.search.v5_x;
+
+import java.util.Collection;
+import java.util.Collections;
+
+import javax.annotation.Nullable;
+
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation;
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.hibernate.search.HibernateSearchConstants;
+import co.elastic.apm.agent.hibernate.search.HibernateSearchHelper;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.NamedElement;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.hibernate.search.query.hibernate.impl.FullTextQueryImpl;
+
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.nameContains;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+
+public class HibernateSearch5Instrumentation extends ElasticApmInstrumentation {
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return HibernateSearch5ExecuteAdvice.class;
+    }
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return nameContains("Query");
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return not(isInterface())
+            .and(hasSuperType(named("org.hibernate.search.FullTextQuery")));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("list")
+            .or(named("scroll"))
+            .or(named("iterate"));
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Collections.singleton(HibernateSearchConstants.HIBERNATE_SEARCH_ORM_TYPE);
+    }
+
+    @VisibleForAdvice
+    public static class HibernateSearch5ExecuteAdvice {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void onBeforeExecute(@Advice.This FullTextQueryImpl query,
```

Comment:
Regarding the question in https://github.com/elastic/apm-agent-java/pull/685#discussion_r298583471- the way to get the method name is by using the `@Advice.Origin` annotation, two ways to get the method name: the explicit way:
```
@Advice.Origin Method method
``` 
which gives you a `java.lang.reflect.Method` from which you can get the name through `getName`, and the utility shortcut:
```
@Advice.Origin("#m") String methodName
```

But maybe the `@SimpleMethodSignature` is actually better? What would you prefer as a user?

Reply:
I would be interested in the names of my own methods. So speaking for the unit tests I would like to get the class and method name of the unit test class.

Reply:
This means going over the stack and search for those, not something we normally do. The way for a user to relate own methods to an instrumented framework is through the stack trace we capture for spans. If you set `application_packages` you would see which of your methods invoked the traced method.

For now, let's go with the method name appended as a postfix to the general prefix as follows: `Hibernate-Search list()` (with a dash in the prefix to separate from the method and with parentheses in the method to indicate this is the instrumented method). See how to get the method name in [this comment](https://github.com/elastic/apm-agent-java/pull/685#discussion_r298821590).

Reply:
That is fine for me. Changed according to the suggestion.